File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int size = 0;
        ListNode tmp = head;
        while(tmp != null){
            tmp = tmp.next;
            size++;
        }
        
        int prevPos = size - n - 1;
        if(prevPos+1 == 0) return head.next;
        
        tmp = head;
        while(prevPos > 0){
            tmp = tmp.next;
            prevPos--;
        }
        
        
        tmp.next = tmp.next.next;
        return head;
    }
}

File: remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun removeNthFromEnd(head: ListNode?, n: Int): ListNode? {
        
        var dummy = ListNode(-1)
        dummy.next = head
        
        var tmp = dummy.next
        var size = 0
        
        while(tmp != null){
            tmp = tmp.next
            size++
        }
        
        size -= n
        
        tmp = dummy
        while(size > 0){
            size--
            tmp = tmp.next
        }
        
        tmp.next = tmp.next.next
        
        return dummy.next
    }
}